tasks/AISTask.cpp

```diff
@@ -36,73 +36,115 @@ bool AISTask::configureHook()
+    m_AIS_stats.received_messages++;
     switch (msg->type()) {
         case ais::message_id::position_report_class_a: {
             auto msg01 = ais::message_cast<ais::message_01>(msg);
-            _positions.write(AIS::getPosition(*msg01));
+            auto position = AIS::getPosition(*msg01);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_assigned_schedule: {
             auto msg02 = ais::message_cast<ais::message_02>(msg);
-            _positions.write(AIS::getPosition(*msg02));
+            auto position = AIS::getPosition(*msg02);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::position_report_class_a_response_to_interrogation: {
             auto msg03 = ais::message_cast<ais::message_03>(msg);
-            _positions.write(AIS::getPosition(*msg03));
+            auto position = AIS::getPosition(*msg03);
+            processPositionReport(position, position.mmsi);
             break;
         }
         case ais::message_id::static_and_voyage_related_data: {
             auto msg05 = ais::message_cast<ais::message_05>(msg);
-            _vessels_information.write(AIS::getVesselInformation(*msg05));
-            _voyages_information.write(AIS::getVoyageInformation(*msg05));
+            auto vesselInformation = AIS::getVesselInformation(*msg05);
+            auto voyageInformation = AIS::getVoyageInformation(*msg05);
+            updateKnownVessels(vesselInformation);
+            _vessels_information.write(vesselInformation);
+            _voyages_information.write(voyageInformation);
             break;
         }
         default:
-            mAISStats.ignored_messages++;
+            m_AIS_stats.ignored_messages++;
             break;
     }
     return true;
 }
+void AISTask::processPositionReport(std::optional<ais_base::Position> position, int mmsi)
+{
+    if (!m_use_sensor_offset_correction) {
+        _positions.write(position.value());
+        return;
+    }
+
+    auto vessel = getCorrespondingVesselInfo(mmsi);
+    if (vessel.has_value()) {
+        auto corrected_position = AIS::applyPositionCorrection(position.value(),
+            vessel->reference_position,
+            m_UTM_converter);
+
+        position.value().latitude = corrected_position.latitude;
+        position.value().longitude = corrected_position.longitude;
+        position.value().correction_status = corrected_position.correction_status;
+    }
+    _positions.write(position.value());
+}
+std::optional<ais_base::VesselInformation> AISTask::getCorrespondingVesselInfo(int mmsi)
+{
+    if (m_vessels.find(mmsi) != m_vessels.end()) {
+        return m_vessels.at(mmsi);
+    }
+
+    return {};
+}
+void AISTask::updateKnownVessels(ais_base::VesselInformation info)
+{
+    if (m_vessels.find(info.mmsi) == m_vessels.end()) {
+        m_vessels[info.mmsi] = info;
+    }
```

Comment:
Correct me if I am wrong, is there any harm in always updating the known vessels list with the latest data? Is there anything missing from a vessel information message that would be an issue? I dont think there is